### ISSUE TITLE
Add IME-like input submit functionality

### ISSUE THREAD

**Issue by r44r1** [Clam-/TraditionalT9#22]
Hi,

I'm using Kyocera flip running Android 5.1 together with your keyboard.

The issue is that in some input areas text can be typed but don't accepted by application. I am facing it primarly with standard apps (inc. lock screen), but in some external too - mainly in Material-themed. The only workaround is to switch to native IME keyboard and confirm previously entered text. But this is very annoying because in android 5 keyboard can't be switched in-place.

Is it possible to add IME submit event action to layout, maybe as a variant for zero key (next to newline)?

**Comment by Clam-** [comment]
Hi,

By default, long pressing the pound/hash key should trigger the "default editor action":
https://github.com/Clam-/TraditionalT9/blob/master/src/org/nyanya/android/traditionalt9/TraditionalT9.java#L645
https://developer.android.com/reference/android/inputmethodservice/InputMethodService.html#sendDefaultEditorAction(boolean)

Note that if the editor has special handling for enter (as in it's possible or even desired to put a new line), it will perform that action instead. But if there's no "special" action required for the enter key, the editor will perform it's default action (send, submit, run, etc). If the editor has not defined a default action, then there's not much I can do unless there's another API I need to check for it.

Please let me know how this goes and if there's any extra information that might be useful.

**Comment by r44r1** [comment]
Longpress doesn't help, it just changes input mode to digits. I use non-T9 mode as default.

Based on info you provided, I suppose that iWnn IME (default keyboard) does some API call after pressing "done" in input field, and framework handles this call and transform it to submit action. I'll try to catch it by DDMS now.

**Comment by r44r1** [comment]
So, I gather up some logs.

Scenario was the same for both keyboards: to click on button which opened text edit area, input some text and click on input submit

My logcat was too overflooded, so I apply filter:

tag:^(?!StatusBar.*|VM_BMS|SurfaceFlinger|TelephonyIcons|ImageView|ServiceManager|Connectivity*|PhoneApp|Location*|EmojiDrawable|QC_RIL*|Qcril*))

I don't think any of this could be tied with text input.

Logs itself: 

<details><summary>T9 with email app, HASH as submit</summary><p>
```
05-06 18:14:59.161: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_DPAD_CENTER, scanCode=28, metaState=0, flags=0x8, repeatCount=0, eventTime=42077660, downTime=42077660, deviceId=8, source=0x301 } repeat?: 0 long-time: false
05-06 18:14:59.171: V/PC mail(28074): [KcNfpActivity]notifyKeyCenter() Not override key event
05-06 18:14:59.251: I/Timeline(28074): Timeline: Activity_launch_request id:com.android.email time:42077757
05-06 18:14:59.251: I/ActivityManager(865): START u0 {cmp=com.android.email/com.android.mail.ui.InputTextActivity (has extras)} from uid 10049 on display 0
05-06 18:14:59.261: V/WindowManager(865): addAppToken: AppWindowToken{1c6a8276 token=Token{33857111 ActivityRecord{1230dd38 u0 com.android.email/com.android.mail.ui.InputTextActivity t34212}}} to stack=1 task=34212 at 2
05-06 18:14:59.311: V/PC mail(28074): [InputTextActivity]onCreate() IN inputType = 131153
05-06 18:14:59.311: I/ViewRootImpl(28074): CPU Rendering VSync enable = true
05-06 18:14:59.321: V/WindowManager(865): Adding window Window{8b294e4 u0 com.android.email/com.android.mail.ui.InputTextActivity} at 11 of 16 (after Window{1e03ea20 u0 com.android.email/com.android.email.activity.ComposeActivityEmail})
05-06 18:14:59.331: D/doLineRunnable(28074): do
05-06 18:14:59.331: D/doLineRunnable(28074): do
05-06 18:14:59.341: V/PhoneStatusBar(2705): setLightsOn(true)
05-06 18:14:59.391: I/Timeline(28074): Timeline: Activity_idle id: android.os.BinderProxy@3a21777d time:42077894
05-06 18:14:59.411: I/ActivityManager(865): Displayed com.android.email/com.android.mail.ui.InputTextActivity: +135ms
05-06 18:14:59.431: I/Timeline(865): Timeline: Activity_windows_visible id: ActivityRecord{1230dd38 u0 com.android.email/com.android.mail.ui.InputTextActivity t34212} time:42077938
05-06 18:14:59.451: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:00.051: W/ResourceType(3169): Failure getting entry for 0x7f020120 (t=1 e=288) (error -75)
05-06 18:15:01.531: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_5, scanCode=6, metaState=0, flags=0x8, repeatCount=0, eventTime=42080030, downTime=42080030, deviceId=8, source=0x301 } repeat?: 0 long-time: false
05-06 18:15:01.531: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:01.541: I/ViewRootImpl(2836): CPU Rendering VSync enable = true
05-06 18:15:01.551: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:01.551: D/doLineRunnable(2836): do
05-06 18:15:02.041: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_6, scanCode=7, metaState=0, flags=0x8, repeatCount=0, eventTime=42080540, downTime=42080540, deviceId=8, source=0x301 } repeat?: 0 long-time: false
05-06 18:15:02.041: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:02.041: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:04.711: D/T9.onEvaluateInputViewShown(2836): whatis

**SUMBIT GOES NEXT - LONG-PRESS HASH KEY**

05-06 18:15:04.711: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=0, eventTime=42083210, downTime=42083210, deviceId=8, source=0x301 } repeat?: 0 long-time: false
05-06 18:15:04.711: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.211: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.211: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x88, repeatCount=1, eventTime=42083711, downTime=42083210, deviceId=8, source=0x301 } repeat?: 1 long-time: true
05-06 18:15:05.211: I/NotificationService(865): cancelToast pkg=org.nyanya.android.traditionalt9 callback=android.app.ITransientNotification$Stub$Proxy@19320468
05-06 18:15:05.221: I/ViewRootImpl(2836): CPU Rendering VSync enable = true
05-06 18:15:05.271: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=2, eventTime=42083762, downTime=42083210, deviceId=8, source=0x301 } repeat?: 2 long-time: false
05-06 18:15:05.271: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.311: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.311: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=3, eventTime=42083813, downTime=42083210, deviceId=8, source=0x301 } repeat?: 3 long-time: false
05-06 18:15:05.371: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=4, eventTime=42083863, downTime=42083210, deviceId=8, source=0x301 } repeat?: 4 long-time: false
05-06 18:15:05.371: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.421: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=5, eventTime=42083913, downTime=42083210, deviceId=8, source=0x301 } repeat?: 5 long-time: false
05-06 18:15:05.421: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.461: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.461: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=6, eventTime=42083963, downTime=42083210, deviceId=8, source=0x301 } repeat?: 6 long-time: false
05-06 18:15:05.521: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.521: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=7, eventTime=42084014, downTime=42083210, deviceId=8, source=0x301 } repeat?: 7 long-time: false
05-06 18:15:05.571: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=8, eventTime=42084065, downTime=42083210, deviceId=8, source=0x301 } repeat?: 8 long-time: false
05-06 18:15:05.571: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.611: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.611: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=9, eventTime=42084116, downTime=42083210, deviceId=8, source=0x301 } repeat?: 9 long-time: false
05-06 18:15:05.661: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.671: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=10, eventTime=42084167, downTime=42083210, deviceId=8, source=0x301 } repeat?: 10 long-time: false
05-06 18:15:05.721: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=11, eventTime=42084218, downTime=42083210, deviceId=8, source=0x301 } repeat?: 11 long-time: false
05-06 18:15:05.721: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.771: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=12, eventTime=42084268, downTime=42083210, deviceId=8, source=0x301 } repeat?: 12 long-time: false
05-06 18:15:05.771: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.821: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.821: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=13, eventTime=42084318, downTime=42083210, deviceId=8, source=0x301 } repeat?: 13 long-time: false
05-06 18:15:05.871: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=14, eventTime=42084369, downTime=42083210, deviceId=8, source=0x301 } repeat?: 14 long-time: false
05-06 18:15:05.871: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.921: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=15, eventTime=42084419, downTime=42083210, deviceId=8, source=0x301 } repeat?: 15 long-time: false
05-06 18:15:05.921: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:05.971: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=16, eventTime=42084471, downTime=42083210, deviceId=8, source=0x301 } repeat?: 16 long-time: false
05-06 18:15:05.971: D/T9.onEvaluateInputViewShown(2836): whatis
05-06 18:15:06.031: D/onKeyDown(2836): Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_POUND, scanCode=253, metaState=0, flags=0x8, repeatCount=17, eventTime=42084521, downTime=42083210, deviceId=8, source=0x301 } repeat?: 17 long-time: false
05-06 18:15:06.031: D/T9.onEvaluateInputViewShown(2836): whatis
```
</p></details>

**Comment by Clam-** [comment]
Long press Pound should not change the input mode to digits, it either inputs return, or it does editor action. The editor might be handling it weird.
I'm afraid the logs don't help. Can you provide screenshots of the interfaces you are talking about? Specifically the iWNN "submit". Maybe a video would make it more clear.

Sorry I cannot help with the work you have done so far.

**Comment by r44r1** [comment]
For hash is my miss, long press changes language. Input mode is changed by double press.

Not a problem, I've maked a bunch of gifs instead: https://imgur.com/a/MDdvg12

I don't make a separate gif for t9+lock+hash as is acts same as t9+lock+center

**Comment by r44r1** [comment]
It's true, Smiley window opens by Star (by short press, long causes dictionary add menu to appears). 
But when multilingual support is disabled (only ENGLISH box is checked in settings), it also shown by long-press Hash. I checked two times and it always happen. Have no idea why.

No, I do not make any custom remaps. I can try to check which scan codes sent in each case tomorrow.

**Comment by Clam-** [comment]
Wow, your buttons are producing very weird actions. You can see what each of the long press actions is supposed to do here:
https://github.com/Clam-/TraditionalT9/blob/4b995fa58261d7561daaf0950a797386702e79ec/src/org/nyanya/android/traditionalt9/TraditionalT9.java#L633
(Note: Soft Left and Soft Right should be the button above the camera and the speaker button on your phone)

Okay, after looking at the GIFs I think there's something different going on.
The submit button in the GIF looks like it's encouraging you to press DPAD center. When T9 input is active, there is supposed to be a input UI that displays. On your device it doesn't seem to show. DPAD Center is supposed to turn the input UI off, and allow you to send DPAD Center actions back to the application.

When you don't see the T9 UI, DPAD Center should be sent directly to the application, but that might not be occurring in your case. Hard to say how to debug this one.
You could try a version before key remapping was implemented but I don't think that will help:
https://nyanya.org/android/old/TraditionalT9-git-1b15cd0.apk
Note this is an old version and is missing features and bug fixes.

**Comment by r44r1** [comment]
But with this one, submit starts to work! Many thanks.

Softkeys does not, but if I get you right - it's just enough to change scancode expected and rebuild app?

**Comment by Clam-** [comment]
So you do see input UI? Can you please show a GIF of the UPD scenario?

Can I also confirm what exact file you installed previously?

Thanks.

**Comment by Clam-** [comment]
Can you please show a GIF of what the SMS app looks like when it works correctly?

You can get the details of the APKs you have if you look at the details of the installed App in the "Installed" or "downloaded" section of your application manager on your device.
I think you may have been using old or custom modified APKs.

**Comment by r44r1** [comment]
It's shown on the last gif from my first link. Direct: https://i.imgur.com/q9K96ZF.gif

Unfortunately, I've uninstalled old APK prior to installing your build. I can't definitely say there I get it, but most probably I use [this build](http://j-phone.ru/sw/ruskey/t9.apk)

Also, I found that on build you provided nput UI starts to show after selecting phone no. from contact in phonebook, making it unable to call via DPAD press. I do not face it on old one.
